block_massaction.php

```diff
@@ -138,6 +138,9 @@ public function get_content(): stdClass {
                     && has_capability('moodle/restore:restoretargetimport', $context)) {
                 $actionicons['duplicate'] = 't/copy';
             }
+            if (has_capability('moodle/backup:backuptargetimport', $context)) {
```

Comment:
Should only show a duplicatetocourse icon if user has ability to restore into another course.  
maybe change to: 

// Get courses that are eligible for import/restore. Filter out siteid and current course.
$siteid = SITEID; // Cannot pass SITEID into filter.
$eligiblecourses = array_filter(get_user_capability_course('moodle/restore:restoretargetimport', null,
                                                           true, '', '', 3),
                                function($obj) use ($COURSE, $siteid) {
                                    if ($obj->id == $siteid || $obj->id == $COURSE->id) {
                                        return false;
                                    }
                                    return true;
                                });
if (has_capability('moodle/backup:backuptargetimport', $context) && count($eligiblecourses) >= 1) {
    $actionicons['duplicatetocourse'] = 't/copy';
}

Reply:
if used, will require a test change.  Two courses and enrollments will be required to see the new tool.  Might keep the 1 course setup to test that the link should NOT show if only 1 course enrollment exists with the correct capabilities

Reply:
Hi @Syxton, thank you for your review.

I already thought about this one, but decided against it. For normal instances this probably would not be a big deal, but our instance has about 400 000 courses, thus we have a big problem calling `get_user_capability_course`, especially on every call of the course page when editing. I'm not sure, if the limit is really helping there. As far as I can tell from the moodle core code there will be still an at least partial join of the `course` and `context` table which in general we want to avoid wherever possible :-)

That's why I decided to only check for `moodle/backup:backuptargetimport` for determining if I want to show the option to a user. If a user does not have enough courses to duplicate to, he will see an empty dropdown in the next page and will be able to just hit "Cancel". That's for sure only the 2nd or 3rd best option from a UX point of view, but avoiding long loading times on course page in editing mode and reducing load onto our instance was more important to me at this point.

Do you have an opinion on that? Especially about the influence of the limit param?

Reply:
After looking more I see what are saying.  the limit is applied after the join, so it doesn't improve efficiencing of the query.  the solution as is will be fine then.  As you said, it is a fairly minor case as well.